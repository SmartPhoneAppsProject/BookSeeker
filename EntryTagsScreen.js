import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  ActivityIndicator,
  FlatList,
} from 'react-native';
import {
  MaterialCommunityIcons,
  Octicons
} from '@expo/vector-icons';
import {
  Input,
  Button,
  List,
  ListItem,
} from 'react-native-elements';

import {
  getTags,
  tagLinkBook,
} from "./networking";
import { icon } from "./icons";

export default class EntryTagsScreen extends Component {
  static navigationOptions = {
    title: 'タグ登録',
  };

  constructor(props) {
    super(props);
    this.state = {
      tagText: '',
      appState: 'isLoading', // or error or success
      tags: [],
      chosenIds: [],
      updated: true,
    };
  }

  componentDidMount() {
    getTags()
      .then((tags) => {
        for (let tag of tags) {
          tag.chosen = false;
        }
        this.setState({
          appStatue: 'success',
          tags,
        });
      })
      .catch(error => {
        console.warn(error)
      });
  }

  renderTagsList = () => {
    return (
      <List containerStyle={styles.listContainer}>
        <FlatList
          keyExtractor={(item) => item.id}
          data={this.state.tags}
          extraData={this.state.updated}
          renderItem={this._renderItem}
        />
      </List>
    );
  };

  _renderItem = ({ item, index }) => {
    const status = item.chosen
      ? <MaterialCommunityIcons name='check-circle-outline' size={25} color='#2e8b57'/>
      : <View/>;

    return (
      <ListItem
        onPress={() => this.changeTagChosen(item, index)}
        title={` ${item.name}`}
        leftIcon={
          icon(item.name)
        }
        badge={{ element: status }}
        hideChevron={true}
      />
    );
  };

  //todo indexから要素のidに変更
  changeTagChosen = (tag, itemIndex) => {
    let { tags, chosenIds } = this.state;

    tags[itemIndex].chosen = !tags[itemIndex].chosen;

    if (tags[itemIndex].chosen) {
      chosenIds.push(tag.id);
    } else {
      const searchIndex = tags.indexOf(tag.id);
      chosenIds.splice(searchIndex, 1);
    }

    console.log(tag);
    console.log(this.state.chosenIds);

    this.setState({
      tags,
      updated: !this.state.updated, //re-render ListView
      chosenIds,
    });
  };

  renderButton = () => {
    return (
      <Button
        containerStyle={styles.buttonContainer}
        buttonStyle={styles.formButton}
        text='submit'
        onPress={this.buttonOnPress}
        iconRight={true}
        icon={
          <MaterialCommunityIcons
            name='arrow-right'
            size={15}
            color='#ffffff'
          />
        }
      />
    );
  };

  buttonOnPress = () => {
    const { book } = this.props.navigation.state.params;
    console.log('book id: ' + book.id);

    for (let id of this.state.chosenIds) {
      this.tagsAssociateToBook(id);
    }
  };

  tagsAssociateToBook = (tagId) => {
    const { book } = this.props.navigation.state.params;
    console.log(book.id);

    const json = JSON.stringify({
      book_id: book.id,
      tag_id: tagId
    });

    tagLinkBook(json)
      .then(response => response.json())
      .then(responseJson => {
        console.log(responseJson);
      })
      .catch(error => {
        console.warn(error);
        tagLinkBook(json)
          .then(response => response.json())
          .then(responseJson => {
            console.log(responseJson);
          })
          .catch(error => console.error(error));
      });
  };

  render() {
    const tags = this.renderTagsList();
    const button = this.renderButton();
    if (this.state.appState == 'success') {
      return (
        <View style={styles.isLoading}>
          <ActivityIndicator/>
        </View>
      );
    }
    return (
      <View style={styles.container}>
        {tags}
        {button}
      </View>
    );
  }
}

const
  styles = StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: '#ffffff',
    },
    isLoading: {
      flex: 1,
      paddingTop: 20
    },
    listContainer: {
      flex: 9,
      marginTop: 0,
    },
    buttonContainer: {
      flex: 1,
    },
    formButton: {
      backgroundColor: '#c0c0c0',
      borderWidth: 0,
      borderRadius: 20,
    },
  });