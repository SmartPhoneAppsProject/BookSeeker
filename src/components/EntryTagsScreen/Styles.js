import { StyleSheet } from 'react-native';

export const index = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#ffffff',
  },
  isLoading: {
    flex: 1,
    paddingTop: 20,
  },
  listContainer: {
    flex: 9,
    marginTop: 0,
  },
  listItemContainer: {
    margin: 10,
  },
  buttonContainer: {
    flex: 1,
  },
  formButton: {
    backgroundColor: '#c0c0c0',
    borderWidth: 0,
    borderRadius: 20,
  },
});
