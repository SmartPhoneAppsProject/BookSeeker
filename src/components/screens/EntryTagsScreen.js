import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  ActivityIndicator,
  FlatList,
} from 'react-native';
import {
  Button,
  List,
  ListItem,
} from 'react-native-elements';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import {
  getTags,
  tagLinkBook,
} from '../../utils/Network';
import { icon } from '../../utils/Icons';
import { navigate } from '../../utils/NavigationService';

export default class EntryTagsScreen extends Component {
  static navigationOptions = {
    title: 'タグ登録',
  };

  constructor(props) {
    super(props);
    this.state = {
      appState: 'isLoading', // or error or success
      tags: [],
      chosenIds: [],
      updated: true,
    };
  }

  componentDidMount() {
    getTags()
      .then((tags) => {
        for (let tag of tags) {
          tag.chosen = false;
        }
        this.setState({
          tags,
        });
      })
      .catch((error) => {
        console.warn(error);
        getTags()
          .then((tags) => {
            for (let tag of tags) {
              tag.chosen = false;
            }
            this.setState({
              tags,
            });
          })
          .catch(e => console.error(e));
      });
  }

  changeTagChosen = (tag, itemIndex) => {
    let { tags, chosenIds } = this.state;

    tags[itemIndex].chosen = !tags[itemIndex].chosen;

    if (tags[itemIndex].chosen) {
      chosenIds.push(tag.id);
    } else {
      const searchIndex = tags.indexOf(tag.id);
      chosenIds.splice(searchIndex, 1);
    }

    this.setState({
      tags,
      updated: !this.state.updated, // re-render BookList
      chosenIds,
    });
  };

  buttonOnPress = () => {
    for (let id of this.state.chosenIds) {
      this.tagAssociateToBook(id);
    }
  };

  tagAssociateToBook = (tagId) => {
    const { book } = this.props.navigation.state.params;
    console.warn(book.id);

    const json = JSON.stringify({
      book_id: book.id,
      tag_id: tagId,
    });

    tagLinkBook(json)
      .then(response => response.json())
      .then((responseJson) => {
        console.log(responseJson);
      })
      .catch((error) => {
        console.warn(error);
        tagLinkBook(json)
          .then(response => response.json())
          .then((responseJson) => {
            console.log(responseJson);
          })
          .catch(e => console.error(e));
      });

    navigate('Home');
  };

  renderTagsList = () => (
    <List containerStyle={styles.listContainer}>
      <FlatList
        keyExtractor={item => item.id}
        data={this.state.tags}
        extraData={this.state.updated}
        renderItem={this.renderListItem}
      />
    </List>
  );

  renderListItem = ({ item, index }) => {
    const status = item.chosen
      ? <MaterialCommunityIcons name="check-circle-outline" size={25} color="#2e8b57" />
      : <View />;

    return (
      <ListItem
        containerStyle={styles.listItemContainer}
        onPress={() => this.changeTagChosen(item, index)}
        title={` ${item.name}`}
        leftIcon={
          icon(item.name)
        }
        badge={{ element: status }}
        hideChevron
      />
    );
  };

  renderButton = () => (
    <Button
      containerStyle={styles.buttonContainer}
      buttonStyle={styles.formButton}
      title="submit"
      onPress={this.buttonOnPress}
      iconRight
      icon={
        <MaterialCommunityIcons
          name="arrow-right"
          size={15}
          color="#ffffff"
        />
      }
    />
  );

  render() {
    const tags = this.renderTagsList();
    const button = this.renderButton();
    if (this.state.appState === 'success') {
      return (
        <View style={styles.isLoading}>
          <ActivityIndicator />
        </View>
      );
    }
    return (
      <View style={styles.container}>
        {tags}
        {button}
      </View>
    );
  }
}

const
  styles = StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: '#ffffff',
    },
    isLoading: {
      flex: 1,
      paddingTop: 20,
    },
    listContainer: {
      flex: 9,
      marginTop: 0,
    },
    listItemContainer: {
      margin: 10,
    },
    buttonContainer: {
      flex: 1,
    },
    formButton: {
      backgroundColor: '#c0c0c0',
      borderWidth: 0,
      borderRadius: 20,
    },
  });
