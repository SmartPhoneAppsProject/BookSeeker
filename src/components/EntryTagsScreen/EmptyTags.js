import React from 'react';
import {
  Text,
  View,
} from 'react-native';

import { EmptyTags as styles } from './Styles';

const EmptyTags = () => (
  <View style={styles.container}>
    <Text style={styles.text}>データを取得できませんでした</Text>
  </View>
);

export default EmptyTags;
