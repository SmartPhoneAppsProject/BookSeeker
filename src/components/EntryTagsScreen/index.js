import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  ActivityIndicator,
  FlatList,
} from 'react-native';
import {
  Button,
  ListItem,
} from 'react-native-elements';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import { index as styles } from './Styles';
import { icon } from '../../utils/Icons';

export default class EntryTagsScreen extends Component {
  componentDidMount() {
    this.props.getAllTags();
  }

  buttonOnPress = () => {
    const {
      title,
      published,
      isbn,
      tags,
      postBook,
      navigation,
    } = this.props;
    const image = this.props.image.base;

    const chosenIds = tags.reduce((accumulator, current) => {
      if (current.chosen) {
        accumulator.push(current.id);
      }
      return accumulator;
    }, []);

    postBook(title, image, published, isbn, chosenIds);
    navigation.navigate('Home');
  };

  renderTagsList = () => {
    const { tags } = this.props;

    return (
      <FlatList
        style={styles.listContainer}
        keyExtractor={item => item.id}
        data={tags}
        extraData={tags}
        renderItem={this.renderListItem}
      />
    );
  }

  renderListItem = ({ item }) => {
    const status = item.chosen
      ? <MaterialCommunityIcons name="check-circle-outline" size={25} color="#2e8b57" />
      : <View />;

    return (
      <ListItem
        containerStyle={styles.listItemContainer}
        onPress={() => this.props.toggleChosenFromId(item.id)}
        title={` ${item.name}`}
        leftIcon={
          icon(item.name)
        }
        badge={{ element: status }}
        hideChevron
      />
    );
  };

  renderButton = () => (
    <Button
      containerStyle={styles.buttonContainer}
      buttonStyle={styles.formButton}
      title="submit"
      onPress={this.buttonOnPress}
      iconRight
      icon={
        <MaterialCommunityIcons
          name="arrow-right"
          size={15}
          color="#ffffff"
        />
      }
    />
  );

  render() {
    const { isLoading } = this.props;
    const tags = this.renderTagsList();
    const button = this.renderButton();
    if (isLoading) {
      return (
        <View style={styles.isLoading}>
          <ActivityIndicator />
        </View>
      );
    }
    return (
      <View style={styles.container}>
        {tags}
        {button}
      </View>
    );
  }
}

EntryTagsScreen.navigationOptions = {
  title: 'タグ登録',
};

EntryTagsScreen.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    chosen: PropTypes.bool,
  })).isRequired,
  title: PropTypes.string.isRequired,
  isbn: PropTypes.string.isRequired,
  published: PropTypes.string.isRequired,
  image: PropTypes.shape({
    uri: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    exif: PropTypes.string,
    base64: PropTypes.string,
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  getAllTags: PropTypes.func.isRequired,
  toggleChosenFromId: PropTypes.func.isRequired,
  postBook: PropTypes.func.isRequired,
};

