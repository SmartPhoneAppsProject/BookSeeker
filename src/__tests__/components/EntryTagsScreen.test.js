import React from 'react';
import { shallow } from 'enzyme';
import EntryTagsScreen from '../../components/EntryTagsScreen';
import _books from '../../api/books.json';

describe('<EntryTagsScreen />', () => {
  it('正しくレンダリングされていること', () => {
    const book = _books[0];
    const wrapper = shallow(<EntryTagsScreen
      title={book.title}
      isbn={book.isbn}
      published={book.published}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
