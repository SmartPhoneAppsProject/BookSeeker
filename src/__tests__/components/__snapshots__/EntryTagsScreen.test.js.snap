// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntryTagsScreen /> 正しくレンダリングされていること 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EntryTagsScreen
    image=""
    isbn="9784774184111"
    published="2018-02-09T00:00:00Z"
    title="javascript "
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <List
          containerStyle={
            Object {
              "flex": 9,
              "marginTop": 0,
            }
          }
        >
          <FlatList
            data={Array []}
            disableVirtualization={false}
            extraData={true}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            numColumns={1}
            onEndReachedThreshold={2}
            renderItem={[Function]}
            scrollEventThrottle={50}
            updateCellsBatchingPeriod={50}
            windowSize={21}
          />
        </List>,
        <Button
          TouchableComponent={[Function]}
          buttonStyle={
            Object {
              "backgroundColor": "#c0c0c0",
              "borderRadius": 20,
              "borderWidth": 0,
            }
          }
          clear={false}
          containerStyle={
            Object {
              "flex": 1,
            }
          }
          disabled={false}
          icon={
            <Icon
              allowFontScaling={false}
              color="#ffffff"
              name="arrow-right"
              size={15}
            />
          }
          iconRight={true}
          loadingProps={
            Object {
              "color": "white",
              "size": "small",
            }
          }
          onPress={[Function]}
          title="submit"
        />,
      ],
      "style": Object {
        "backgroundColor": "#ffffff",
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <FlatList
            data={Array []}
            disableVirtualization={false}
            extraData={true}
            horizontal={false}
            initialNumToRender={10}
            keyExtractor={[Function]}
            maxToRenderPerBatch={10}
            numColumns={1}
            onEndReachedThreshold={2}
            renderItem={[Function]}
            scrollEventThrottle={50}
            updateCellsBatchingPeriod={50}
            windowSize={21}
          />,
          "containerStyle": Object {
            "flex": 9,
            "marginTop": 0,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data": Array [],
            "disableVirtualization": false,
            "extraData": true,
            "horizontal": false,
            "initialNumToRender": 10,
            "keyExtractor": [Function],
            "maxToRenderPerBatch": 10,
            "numColumns": 1,
            "onEndReachedThreshold": 2,
            "renderItem": [Function],
            "scrollEventThrottle": 50,
            "updateCellsBatchingPeriod": 50,
            "windowSize": 21,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "TouchableComponent": [Function],
          "buttonStyle": Object {
            "backgroundColor": "#c0c0c0",
            "borderRadius": 20,
            "borderWidth": 0,
          },
          "clear": false,
          "containerStyle": Object {
            "flex": 1,
          },
          "disabled": false,
          "icon": <Icon
            allowFontScaling={false}
            color="#ffffff"
            name="arrow-right"
            size={15}
          />,
          "iconRight": true,
          "loadingProps": Object {
            "color": "white",
            "size": "small",
          },
          "onPress": [Function],
          "title": "submit",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <List
            containerStyle={
              Object {
                "flex": 9,
                "marginTop": 0,
              }
            }
          >
            <FlatList
              data={Array []}
              disableVirtualization={false}
              extraData={true}
              horizontal={false}
              initialNumToRender={10}
              keyExtractor={[Function]}
              maxToRenderPerBatch={10}
              numColumns={1}
              onEndReachedThreshold={2}
              renderItem={[Function]}
              scrollEventThrottle={50}
              updateCellsBatchingPeriod={50}
              windowSize={21}
            />
          </List>,
          <Button
            TouchableComponent={[Function]}
            buttonStyle={
              Object {
                "backgroundColor": "#c0c0c0",
                "borderRadius": 20,
                "borderWidth": 0,
              }
            }
            clear={false}
            containerStyle={
              Object {
                "flex": 1,
              }
            }
            disabled={false}
            icon={
              <Icon
                allowFontScaling={false}
                color="#ffffff"
                name="arrow-right"
                size={15}
              />
            }
            iconRight={true}
            loadingProps={
              Object {
                "color": "white",
                "size": "small",
              }
            }
            onPress={[Function]}
            title="submit"
          />,
        ],
        "style": Object {
          "backgroundColor": "#ffffff",
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <FlatList
              data={Array []}
              disableVirtualization={false}
              extraData={true}
              horizontal={false}
              initialNumToRender={10}
              keyExtractor={[Function]}
              maxToRenderPerBatch={10}
              numColumns={1}
              onEndReachedThreshold={2}
              renderItem={[Function]}
              scrollEventThrottle={50}
              updateCellsBatchingPeriod={50}
              windowSize={21}
            />,
            "containerStyle": Object {
              "flex": 9,
              "marginTop": 0,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "data": Array [],
              "disableVirtualization": false,
              "extraData": true,
              "horizontal": false,
              "initialNumToRender": 10,
              "keyExtractor": [Function],
              "maxToRenderPerBatch": 10,
              "numColumns": 1,
              "onEndReachedThreshold": 2,
              "renderItem": [Function],
              "scrollEventThrottle": 50,
              "updateCellsBatchingPeriod": 50,
              "windowSize": 21,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "TouchableComponent": [Function],
            "buttonStyle": Object {
              "backgroundColor": "#c0c0c0",
              "borderRadius": 20,
              "borderWidth": 0,
            },
            "clear": false,
            "containerStyle": Object {
              "flex": 1,
            },
            "disabled": false,
            "icon": <Icon
              allowFontScaling={false}
              color="#ffffff"
              name="arrow-right"
              size={15}
            />,
            "iconRight": true,
            "loadingProps": Object {
              "color": "white",
              "size": "small",
            },
            "onPress": [Function],
            "title": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
